### PR TITLE
Add Vecchia approximation

### PR DESCRIPTION
**Pull request by samanklesaria**
I thought I might add the Vecchia approximation, as described in https://arxiv.org/pdf/2102.13299.pdf.

### CODE REVIEW

File: src/NearestNeighborsModule.jl [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+function AbstractGPs.posterior(nn::NearestNeighbors, fx::AbstractGPs.FiniteGP, y::AbstractVector)
+	kern = fx.f.kernel
+	F = make_F(fx.x, nn.k, kern)
+	B = make_B(fx.x, nn.k, kern)
+	U = UpperTriangular((I - B)' * inv(sqrt(F)))
+	δ = y - mean(fx)
+	α = U * (U' * δ)
+	C = InvRoot(U)
+	AbstractGPs.PosteriorGP(fx.f, (α=α, C=C, x=fx.x, δ=δ))
```
Comment (willtebbutt) [inline src/NearestNeighborsModule.jl:62]:
```suggestion
	return AbstractGPs.PosteriorGP(fx.f, (α=α, C=C, x=fx.x, δ=δ))
```
Style

File: src/NearestNeighborsModule.jl [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		else
+			ns = pts[max(1, i-k):i-1]
+			ki = kern.(ns, pts[i])
+			push!(vals, prior - dot(ki, kernelmatrix(kern, ns) \ ki))
+		end
+	end
+	Diagonal(vals)
+end
+
+struct NearestNeighbors
```
Comment (willtebbutt) [inline src/NearestNeighborsModule.jl:42]: Could you please add a docstring to explain what this approximation does, what the value of `k` determines, and to provide a reference / references?

File: src/NearestNeighborsModule.jl [hunk 3]
```diff
@@ -0,0 +1,65 @@
+module NearestNeighborsModule
+
+using KernelFunctions, LinearAlgebra, SparseArrays, AbstractGPs
+
+function make_B(pts::AbstractVector{T}, k::Int, kern::Kernel) where {T}
```
Comment (willtebbutt) [inline src/NearestNeighborsModule.jl:5]: Could you please add a comment explaining what this function does.

Additionally, is there a more descriptive name possible than `make_B`? As someone who is aware that this approximation exists, but hasn't worked with it previously, I have no idea what `make_B` might mean haha

File: src/NearestNeighborsModule.jl [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		start_ix = max(i-k, 1)
+		col_ixs = start_ix:(start_ix + length(row) - 1)
+		append!(js, col_ixs)
+		append!(is, fill(i, length(col_ixs)))
+		append!(vals, row)
+	end
+	sparse(is, js, vals, n, n)
+end
+
+function make_F(pts::AbstractVector{T}, k::Int, kern::Kernel) where {T}
```
Comment (willtebbutt) [inline src/NearestNeighborsModule.jl:26]: Same comment as `make_F`.

File: test/NearestNeighborsModule.jl [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    @testset "Using nearest neighbors approximates the exact GP" begin
+        opt_pred = mean_and_cov(posterior(NearestNeighbors(3), fx, y)(x2))
+        pred = mean_and_cov(posterior(fx, y)(x2))
+        for i in 1:2
+            @test all(isapprox.(opt_pred[i], pred[i]; atol=1e-1))
+        end
+    end
+
+    @testset "Zygote can take gradients of the logpdf" begin
+        function objective(lengthscale, var)
```
Comment (willtebbutt) [inline test/NearestNeighborsModule.jl:26]: Does this function use the Vecchia approximation at all? It looks to me like it just does exact inference.

**Review by willtebbutt — COMMENTED** [review]
Thanks for opening this @samanklesaria . Will be good to have!

Does the Vecchia approximation provide an approximation to the log marginal likelihood? Would be great to have that available if possible.

File: test/NearestNeighborsModule.jl (hunk 5 above)
Comment (samanklesaria) [reply test/NearestNeighborsModule.jl:32]: You're quite correct- I entirely forgot to include the Vecchia approximation here. I should have just pushed a fix for this.

File: src/NearestNeighborsModule.jl (hunk 3 above)
Comment (samanklesaria) [reply src/NearestNeighborsModule.jl:15]: An excellent point! I have added a little documentation, which hopefully should make things clearer.

**Comment by willtebbutt** [comment]
@samanklesaria please do let me know when this is ready for another review.

**Comment by samanklesaria** [comment]
@willtebbutt all set for another review! I don't know why all the CI jobs are failing, but it seems to be unrelated to the code here.

File: src/NearestNeighborsModule.jl [hunk 6]
```diff
@@ -0,0 +1,116 @@
+module NearestNeighborsModule
+using ..API
+using ChainRulesCore
+using KernelFunctions, LinearAlgebra, SparseArrays, AbstractGPs
+
+"""
+Constructs the matrix ``B`` for which ``f = Bf + \epsilon`` were ``f``
```
Comment (willtebbutt) [inline src/NearestNeighborsModule.jl:7]:
```suggestion
Constructs the matrix ``B`` for which ``f = Bf + \epsilon`` where ``f``
```

**Review by willtebbutt — COMMENTED** [review]
Thanks for improving the PR. I've made a few cosmetic remarks, but I think we're nearly ready to go now.

File: src/NearestNeighborsModule.jl [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+function make_row(kern::Kernel, ns::AbstractVector{T}, p::T) where {T}
+    kernelmatrix(kern,ns) \ kern.(ns, p)
+end
+
+function make_js(rows, k)
+    [begin
+        start_ix = max(i-k, 1)
+        start_ix:(start_ix + length(row) - 1)
+    end for (row, i) in zip(rows, 2:(length(rows)+1))]
```
[the remark this answers is not in the document]
Comment (samanklesaria) [reply src/NearestNeighborsModule.jl:35]: Sounds good!

**Comment by willtebbutt** [comment]
Ah, sorry, you'll need to import `log2\pi` from IrrationalConstants.jl -- that's causing test failures in your tests at the minute.

**Comment by samanklesaria** [comment]
My gradient tests are failing now- investigating it.

**Comment by samanklesaria** [comment]
Everything seems to be working! Thanks for the suggestions.

**Comment by willtebbutt** [comment]
Wonderful -- just a couple of formatting suggestions to fix, then we can merge.

**Comment by samanklesaria** [comment]
Are the changes from `JuliaFormatter` enough, or should I fix other formatting issues as well?

**Comment by willtebbutt** [comment]
I'm going to merge as-is if you're happy @samanklesaria  -- the AD issues will need to be addressed in another PR.

**Comment by samanklesaria** [comment]
Sounds good! I can't seem to reproduce the AD issues. Might have been a quirk with an inconsistent `Revise` state.

**Comment by willtebbutt** [comment]
Thanks for bumping the version. I'm going to merge + tag despite the test failures, because AD problems need to be addressed elsewhere.

**Comment by willtebbutt** [comment]
Thanks for the contribution @samanklesaria

**Pull request merged**